src/yang/ietf-tcp.yang

```diff
@@ -190,26 +182,17 @@ module ietf-tcp {
              that can be part of either the client or server.";
         }
         description
-          "Connection related parameters.";
+          "List of TCP connections with their parameters. The list 
+           is modelled as writeable, but implementations may not allow
+           creation of new TCP connections by adding entries to the list.
+           Furthermore, the behavior upon removal is implementation-specific.
+           Implementations may support closing or reseting a TCP connection 
+           upon an operation that removes the entry from the list.";
       }
       description
         "A container of all TCP connections.";
     }
 
-    container server {
-      if-feature server;
-      uses tcps:tcp-server-grouping;
-      description
-        "Definitions of TCP server configuration.";
-    }
-
-    container client {
-      if-feature client;
-      uses tcpc:tcp-client-grouping;
-      description
-        "Definitions of TCP client configuration.";
-    }
-
```

Comment:
I can see the desire to minimize what this model covers. However, if an implementation needs support for TCP keepalives, how would they realize that configuration?

Reply:
There are two options for supporting keep-alives:

(1) Keep-alives are configured in the application, which can directly use the client and/or server models in draft-ietf-netconf-tcp-client-server, e.g., by importing the groupings defined there. In that case, draft-ietf-tcpm-yang-tcp is not needed at all.

(2) A NETCONF client sets the keepalives per connection by writing to the TCP connection list; the keep-alive configuration is imported  for each connection in tcp-common-grouping. This may require write access to the keep-alive parameters int he list, but this is what is in the model already...

The former case corresponds to setting the parameters via the sockets API, the latter corresponds to sysctls.

Makes sense?

Reply:
> There are two options for supporting keep-alives:
> 
> (1) Keep-alives are configured in the application, which can directly use the client and/or server models in draft-ietf-netconf-tcp-client-server, e.g., by importing the groupings defined there. In that case, draft-ietf-tcpm-yang-tcp is not needed at all.

Ok. I am an application. For my application to function, I need to configure TCP keepalives. I do not need to configure anything else, and as such I do not need a YANG model. But since the keepalive is a grouping, I need to come up with a dummy YANG model, import the grouping from draft-ietf-netconf-tcp-client-server, just to configure the keepalive. If there are two separate applications both of which need to configure keepalives, I repeat the process there??

> 
> (2) A NETCONF client sets the keepalives per connection by writing to the TCP connection list in _this_ model; the keep-alive configuration is imported for each connection in tcp-common-grouping. This may require write access to the keep-alive parameters in the list, but this is what is in the model already...

Ok. I am confused now. The diffs suggest that we are removing the TCP client grouping. If we do that, how is keepalive configured?

     +--rw client {client}?
     |  +--rw remote-address    inet:host
     |  +--rw remote-port?      inet:port-number
     |  +--rw local-address?    inet:ip-address
     |  +--rw local-port?       inet:port-number
     **|  +--rw keepalives!
     |     +--rw idle-time         uint16
     |     +--rw max-probes        uint16
     |     +--rw probe-interval    uint16**

> 
> The former case corresponds to setting the parameters via the sockets API, the latter corresponds to sysctls.
> 
> Makes sense?

Reply:
Regarding (1): Yep, that is correct, but that is not different how an application would configure keep-alives today. Whether to set keep-alives would be configured inside the application (e.g., its config file in whatever format). If that config file is moved to YANG, the groupings from draft-ietf-netconf-tcp-client-server could be used. HOWEVER, it seems unlikely that an application would only require keep-alive configuration and not any of the other configuration options offered by the set of YANG models developed in the NETCONF WG, which draft-ietf-netconf-tcp-client-server is part of. In a more realistic case, I would assume that the app would use a larger set of models developed in the NETCONF WG, say, also TLS. I assume that the NETCONF models are applicable in that case and draft-ietf-netconf-tcp-client-server is sufficient in that case.

Regarding (2): No, the TCP client grouping stays inside the TCP connection list. Keep-alives can still be configured per connection in this YANG model, using the common grouping importet from draft-ietf-netconf-tcp-client-server.

Note that IMHO *this* model is about all TCP connections that a network node has, which will typcially be more than one. So, IMHO any connection-specific configuration must inherently be done inside the list of connection, not outside. It does not make sense to have entities for a single client and/or server in this model.

So, my proposal is to remove the single client and server configuration *outside* the connection list only.